### ISSUE TITLE
[Bug] Warning! -> warning: The size of tensor a (4070) must match the size of tensor b (4096) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([4096, 6144])

### ISSUE THREAD

**Issue by lmx760581375** [OpenGVLab/InternVL#550]
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.
- [ ] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.

### Describe the bug

训练的时候，为什么会报如下 warning？是否会影响训练效果？

{'loss': 1.2974, 'grad_norm': 5.156000420281073, 'learning_rate': 1.0810810810810812e-06, 'epoch': 0.02}
  0%|          | 6/3700 [06:25<63:02:44, 61.44s/it]warning: The size of tensor a (4070) must match the size of tensor b (4096) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([4096, 6144])
dynamic ViT batch size: 22, images per sample: 11.0, dynamic token length: 4096
warning: The size of tensor a (4070) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
warning: The size of tensor a (4070) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
[2024-08-30 17:18:45,859] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd_microstep: 7624.94 | bwd_microstep: 20504.66 | bwd_inner_microstep: 20483.91 | bwd_allreduce_microstep: 20.67 | step_microstep: 0.09
dynamic ViT batch size: 2, images per sample: 1.0, dynamic token length: 1458
warning: The size of tensor a (4070) must match the size of tensor b (4096) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([4096, 6144])
warning: The size of tensor a (4326) must match the size of tensor b (5632) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4326, 6144]), vit_embeds.shape=torch.Size([5632, 6144])
warning: The size of tensor a (4070) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
warning: The size of tensor a (5350) must match the size of tensor b (6656) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([5350, 6144]), vit_embeds.shape=torch.Size([6656, 6144])
[2024-08-30 17:19:22,967] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | optimizer_step: 3141.06
[2024-08-30 17:19:22,969] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd_microstep: 6013.96 | bwd_microstep: 24795.71 | bwd_inner_microstep: 24736.25 | bwd_allreduce_microstep: 59.36 | step_microstep: 6289.07
[2024-08-30 17:19:22,969] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd: 13638.88 | bwd: 45300.36 | bwd_inner: 45220.16 | bwd_allreduce: 80.05 | step: 6289.16
{'loss': 0.67, 'grad_norm': 3.433773964762511, 'learning_rate': 1.2612612612612613e-06, 'epoch': 0.02}
  0%|          | 7/3700 [07:30<64:19:28, 62.70s/it]dynamic ViT batch size: 22, images per sample: 11.0, dynamic token length: 4096
warning: The size of tensor a (4064) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4064, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
warning: The size of tensor a (5862) must match the size of tensor b (7168) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([5862, 6144]), vit_embeds.shape=torch.Size([7168, 6144])
[2024-08-30 17:19:52,899] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd_microstep: 8106.76 | bwd_microstep: 21713.66 | bwd_inner_microstep: 21065.60 | bwd_allreduce_microstep: 647.97 | step_microstep: 0.12
dynamic ViT batch size: 2, images per sample: 1.0, dynamic token length: 520
warning: The size of tensor a (4070) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
warning: The size of tensor a (4838) must match the size of tensor b (7168) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4838, 6144]), vit_embeds.shape=torch.Size([7168, 6144])

warning: The size of tensor a (4070) must match the size of tensor b (4096) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([4096, 6144])
warning: The size of tensor a (4326) must match the size of tensor b (5632) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4326, 6144]), vit_embeds.shape=torch.Size([5632, 6144])
warning: The size of tensor a (4070) must match the size of tensor b (5376) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([4070, 6144]), vit_embeds.shape=torch.Size([5376, 6144])
warning: The size of tensor a (5350) must match the size of tensor b (6656) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([5350, 6144]), vit_embeds.shape=torch.Size([6656, 6144])

### Reproduction

 GPUS=${GPUS:-32}
BATCH_SIZE=${BATCH_SIZE:-128}
PER_DEVICE_BATCH_SIZE=${PER_DEVICE_BATCH_SIZE:-2}
GRADIENT_ACC=$((BATCH_SIZE / PER_DEVICE_BATCH_SIZE / GPUS))

OUTPUT_DIR='/InternVL2-26B-sft'
MODEL_NAME_OR_PATH="/apdcephfs_cq10/share_1567347/share_info/llm_models/InternVL2-26B"
DATA_PATH="shell/data/internvl_2_finetune_v3.json"
if [ ! -d "$OUTPUT_DIR" ]; then
  mkdir -p "$OUTPUT_DIR"
fi

nohup deepspeed --hostfile=/tmp/mae_hostfile  internvl/train/internvl_chat_finetune_deepspeed.py \
  --model_name_or_path ${MODEL_NAME_OR_PATH} \
  --conv_style "internlm2-chat" \
  --output_dir ${OUTPUT_DIR} \
  --meta_path ${DATA_PATH} \
  --overwrite_output_dir True \
  --force_image_size 448 \
  --max_dynamic_patch 24 \
  --down_sample_ratio 0.5 \
  --drop_path_rate 0.4 \
  --pad2square False \
  --freeze_llm False \
  --freeze_mlp False \
  --freeze_backbone False \
  --vision_select_layer -1 \
  --use_data_resampling False \
  --dataloader_num_workers 4 \
  --fp16 True \
  --num_train_epochs 10 \
  --per_device_train_batch_size ${PER_DEVICE_BATCH_SIZE} \
  --gradient_accumulation_steps ${GRADIENT_ACC} \
  --evaluation_strategy "no" \
  --save_strategy "epoch" \
  --save_total_limit 10 \
  --learning_rate 2e-5 \
  --weight_decay 0.05 \
  --warmup_ratio 0.03 \
  --lr_scheduler_type "cosine" \
  --logging_steps 1 \
  --max_seq_length 4096 \
  --do_train True \
  --grad_checkpoint True \
  --group_by_length True \
  --dynamic_image_size True \
  --use_thumbnail True \
  --ps_version 'v2' \
  --deepspeed "zero_stage3_config.json" \
  --report_to "tensorboard" > ${OUTPUT_DIR}/training_log.txt 2>&1 &

### Environment

```Shell
Cuda12.2
python310
torch2.2.2
```

**Comment by czczup** [comment]
Same as #185 and #461.

**Comment by whale1008** [comment]
InternVl construct tokens first， then fill vit_embeds.
The former is controlled by --max_seq_length, the latter by --max_dynamic_patch (if --force_image_size is kept)
if input_embeds[selected].shape[0] < vit_embeds.shape[0]，Increasing max_seq_length or lowering max_dynamic_patch can help you.
if input_embeds[selected].shape[0] > vit_embeds.shape[0]，there is probably an <IMG_CONTEXT> in the build prompt, which conflicts with the token placeholder of vit_embed. Replacing or deleting the <IMG_CONTEXT> in the question in the dataset can help you.

**Comment by awakenlee180** [comment]
warning: The size of tensor a (256) must match the size of tensor b (2048) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([256, 896]), vit_embeds.shape=torch.Size([2048, 896])


但是我的max_seq_length=4096 也没超出啊 奇怪

**Comment by czczup** [comment]
> shape=torch.Size

用8192试试看？

**Comment by awakenlee180** [comment]
解决了 多图对话不适合我的数据集
